pulsar/producer_partition.go

```diff
@@ -434,17 +445,79 @@ func (p *partitionProducer) internalFlushCurrentBatch() {
 		return
 	}
 
+	p.queueLock.Lock()
 	p.pendingQueue.Put(&pendingItem{
+		createdAt:    time.Now(),
 		batchData:    batchData,
 		sequenceID:   sequenceID,
 		sendRequests: callbacks,
 	})
+	p.queueLock.Unlock()
+
 	p.cnx.WriteData(batchData)
 }
 
+func (p *partitionProducer) failTimeoutMessages(sendTimeout time.Duration) {
+	t := time.NewTimer(sendTimeout)
+	var state int32
+	for range t.C {
+		// since Closing/Closed state could not be reopen, load and compare is safe
+		state = atomic.LoadInt32(&p.state)
+		if state == producerClosing || state == producerClosed {
+			t.Stop()
+			return
+		}
+
+		p.queueLock.Lock()
+		item := p.pendingQueue.Peek()
+		if item == nil {
+			t.Reset(p.options.SendTimeout)
+			p.queueLock.Unlock()
+			continue
+		}
+
+		pi := item.(*pendingItem)
+		diff := p.options.SendTimeout - time.Since(pi.createdAt)
+		if diff > 0 {
+			// pending messages not timeout yet
+			t.Reset(diff)
+			p.queueLock.Unlock()
+			continue
+		}
+
+		p.log.Infof("Failing %d messages", p.pendingQueue.Size())
+		for p.pendingQueue.Size() > 0 {
+			pi = p.pendingQueue.Poll().(*pendingItem)
+			pi.Lock()
+			for _, i := range pi.sendRequests {
+				sr := i.(*sendRequest)
+				if sr.msg != nil {
+					size := len(sr.msg.Payload)
+					p.publishSemaphore.Release()
+					messagesPending.Dec()
+					bytesPending.Sub(float64(size))
+					publishErrors.Inc()
+					p.log.WithError(errSendTimeout).
+						WithField("size", size).
+						WithField("properties", sr.msg.Properties)
+				}
+				if sr.callback != nil {
+					sr.callback(nil, sr.msg, errSendTimeout)
+				}
+			}
+			buffersPool.Put(pi.batchData)
+			pi.Unlock()
+		}
+		t.Reset(p.options.SendTimeout)
+		p.queueLock.Unlock()
+	}
+}
+
 func (p *partitionProducer) internalFlush(fr *flushRequest) {
 	p.internalFlushCurrentBatch()
 
+	p.queueLock.Lock()
```

Comment:
Instead of having a separate go routine checking for timeouts can the message expiry check be performed here?

I don't think the timeout check needs to be exact? We just need to expires old messages at some point? This could simplify the code and reduces the need for another lock. We can just call this method `failTimeoutMessages` and expire messages if needed.

Reply:
Thanks for review, good idea.
In some case `eventLoop` will be stucked such as producer is reconnecting to broker, 2 ways:
- Check at independent goroutine: `sendTimeout` still effective, periodic check and fail pending messages.
- Check before flush: `sendTimeout` not effective in this case.

If no need to check message timeout precisely, the above case will be ignored.
This is reasonable, since sendTimeout now implemented at **producer level**, no need to be that precise, also avoid the cost of lock.